crates/providers/kv-vault/src/lib.rs

```diff
@@ -1,132 +1,123 @@
-#[derive(Default, Clone, Provider)]
-#[services(KeyValue)]
-struct KvVaultProvider {
+wasmcloud_provider_wit_bindgen::generate!({
+    impl_struct: KvVaultProvider,
+    contract: "wasmcloud:keyvalue",
+    wit_bindgen_cfg: "provider-kv-vault"
+});
+
+/// Redis KV provider implementation which utilizes [Hashicorp Vault](https://developer.hashicorp.com/vault/docs)
+#[derive(Default, Clone)]
+pub struct KvVaultProvider {
     // store redis connections per actor
     actors: std::sync::Arc<RwLock<HashMap<String, RwLock<Client>>>>,
 }
-/// use default implementations of provider message handlers
-impl ProviderDispatch for KvVaultProvider {}
 
-/// Handle provider control commands
-/// put_link (new actor link command), del_link (remove link command), and shutdown
+impl KvVaultProvider {
+    /// Retrieve a client for a given context (determined by actor_id)
+    async fn get_client(&self, ctx: &Context) -> ProviderInvocationResult<Client> {
+        // get the
+        let actor_id = ctx.actor.as_ref().ok_or_else(|| {
+            ProviderInvocationError::Provider("invalid parameter: no actor in request".into())
+        })?;
+        // Clone the existing client for the given actor from the internal hash map
+        let client = self
+            .actors
+            .read()
+            .await
+            .get(actor_id)
+            .ok_or_else(|| {
+                ProviderInvocationError::Provider(format!(
+                    "invalid parameter: actor [{actor_id}] not linked"
+                ))
+            })?
+            .read()
+            .await
+            .clone();
+        Ok(client)
+    }
+}
+
+/// Handle provider control commands, the minimum required of any provider on
+/// a wasmcloud lattice
 #[async_trait]
-impl ProviderHandler for KvVaultProvider {
+impl WasmcloudCapabilityProvider for KvVaultProvider {
     /// Provider should perform any operations needed for a new link,
     /// including setting up per-actor resources, and checking authorization.
     /// If the link is allowed, return true, otherwise return false to deny the link.
     #[instrument(level = "debug", skip(self, ld), fields(actor_id = %ld.actor_id))]
-    async fn put_link(&self, ld: &LinkDefinition) -> RpcResult<bool> {
-        let config = Config::from_values(&ld.values)?;
-        let client = Client::new(config).map_err(to_rpc_err)?;
+    async fn put_link(&self, ld: &LinkDefinition) -> bool {
+        let config = if let Ok(config) =
+            Config::from_values(&HashMap::from_iter(ld.values.clone().into_iter()))
+        {
+            config
+        } else {
+            return false;
+        };
+
+        let client = if let Ok(client) = Client::new(config.clone()) {
+            client
+        } else {
+            return false;
+        };
+
         let mut update_map = self.actors.write().await;
-        info!("adding link for actor");
+        info!("adding link for actor id [{}]", ld.actor_id);
```

Comment:
Is there a reason to prefer embedding the actor ID in the string, over including it as a separate field? These messages aren't super common so I'm not that concerned, but "low-cardinality log messages" has been drilled into my head 🙂